pkg/storage/storage.go

```diff
@@ -163,6 +164,49 @@ func New(cfg *config.Config) (*Storage, error) {
 		return tree.New()
 	}
 
+	// load the total memory of the server
+	vm, err := mem.VirtualMemory()
+	if err != nil {
+		return nil, err
+	}
+
+	interval := time.Duration(10)
+	// start a timer for checking if the memory used by application
+	// is more than 25% of the total memory, if so, trigger eviction
+	// with 10% to every cache
+	go func() {
+		ticker := time.NewTimer(interval * time.Second)
+		defer ticker.Stop()
+
+		for range ticker.C {
+			if s.closing {
```

Comment:
I guess this will cause races.

Reply:
can u help to describe the races?

Reply:
There is a race condition on access to `closing` field, you should acquire `closingMutex` and release it only after `Evict` calls. See `Close`: basically, I think that we could replace the mutex with an `RWMutex`.

Reply:
for range ticker.C {
			if s.closing {
				return
			}

Reply:
there is no closingMutex here

Reply:
actually, i think for the field closing, the mutex is not necessary. the code is weird for using the mutex for closing

Reply:
I'm afraid, without a mutex, there is no other way to avoid the race (without changing too many things): eventually the call will conflict with `Close`.

Reply:
it's a bool variable, i think it's not a problem

Reply:
Regardless of the type, when the same address accessed for read-write concurrently, a race condition occurs. The funny part is that it would be safer to not check `closing` field at all in this particular function, because `Evict` and `Flush` (in `Close`) methods can be called in arbitrary order with no issues :)